### DISCUSSION TITLE
select dont rebuild without setState

### DISCUSSION THREAD

**Discussion by BigGitWorld** [rrousselGit/riverpod#2031] [General]
Hello everyone.
I have an object like as the below:
```
class MainModel {
  String name;
  List<SubModel> products;
  //constructor,...
}
class SubModel {
  String name;
  String price;
  //constructor,...
}
```
I'm going to watch changes of ``products`` property of ``MainModel class`` not whole of the object. So I'm using ``select``.

provider:
```
final myStateProvider = StateNotifierProvider<MyStateNotifierProvider, MainModel>((ref) => MyStateNotifierProvider(ref));
class MyStateNotifierProvider extends StateNotifier<MainModel>{

   //constructor
  
  //this function updates ``products`` property 
  void updateProducts(SubModel s){
      state.products.add(s);
      state = MainModel(name: state.name, products: state.products); //in this line, I'm updating state to the new one
   }
}
```

Watching in the UI layer:
```
class MyView extends ConsumerStatefulWidget {
  const MyView({Key? key}) : super(key: key);

  @override
  ConsumerState<MyView> createState() => _MyViewState();
}

class _MyViewState extends ConsumerState<MyView> {
  @override
  Widget build(BuildContext context) {
   List< SubModel > products = ref.watch(myStateProvider.select((mainModel) {
        //setState(() {});  // (A)
        debugPrint("called");
        return mainModel.producers;
      }));
    return Container();
  }
}
```

But ``MyView`` (in UI layer) will not rebuild, although ``debugPrint`` is called whenever ``products`` property is changed.
If I use ``setState((){})`` (in A), my UI will be rebuild.

what is the problem?

dependencies:
```
freezed: any
flutter_hooks: ^0.18.0
hooks_riverpod: ^1.0.2
```

**Comment by Quijx** [comment]
By doing `state.products.add(s);` in the `updateProducts`, you are modifying the state of the current `MainModel`. When you then assign the `state` a new `MainModel` with the same name and products, you are creating a new instance, but the new instance contains the same products list as the one in `state` before the assignment. When you then select the products, it will try to compare the new state with the old state. But since they are the same instance, the provider will not update.

Riverpod works best with immutable state. Try for example using https://pub.dev/packages/freezed to generate your model.
Then you will have to do:
```dart
   void updateProducts(SubModel s){
      state = state.copyWith(products: [...state.products, s]);
   }
```
